### ISSUE TITLE
[AArch64][SVE] Faster unsigned narrowing to improve the parallelism degree of loop iteration processing

### ISSUE THREAD

**Issue by vfdff** [llvm/llvm-project#64986]
test: https://godbolt.org/z/1bo7TrYWj
```
void pixel_avg(uint8_t *__restrict dst, int i_dst_stride,
               uint8_t *__restrict src1, int i_src1_stride,
               uint8_t *__restrict src2, int i_src2_stride,
               int i_width, int i_height )
{
    for( int y = 0; y < i_height; y++ )
    {
        for( int x = 0; x < i_width; x++ )
            dst[x] = ( src1[x] + src2[x] + 1 ) >> 1;
        dst += i_dst_stride;
        src1 += i_src1_stride;
        src2 += i_src2_stride;
    }
}
```

this may be similar to the new **faster unsigned narrowing** feature in  [gcc12](https://community.arm.com/arm-community-blogs/b/tools-software-ides-blog/posts/gcc-12)
 * **llvm**: compute 8 channels each iteration
 ```
.LBB0_12: // Parent Loop BB0_4 Depth=1
  // => This Inner Loop Header: Depth=2
  ld1b { z0.h }, p1/z, [x2, x18]
  ld1b { z1.h }, p1/z, [x4, x18]
  urhadd z0.h, p1/m, z0.h, z1.h
  st1b { z0.h }, p1, [x0, x18]
  add x18, x18, x16
  cmp x17, x18
  b.ne .LBB0_12
```
 * **gcc**: compute 16 channels each iteration
```
.L5:
  ld1b z31.b, p7/z, [x2, x1]
  ld1b z30.b, p7/z, [x4, x1]
  urhadd z31.b, p6/m, z31.b, z30.b
  st1b z31.b, p7, [x0, x1]
  incb x1
  whilelo p7.b, w1, w6
  b.any .L5
```

**Comment by vfdff** [comment]
* This issue just a **cost model** issue, https://reviews.llvm.org/D158988 
```
LV: Found an estimated cost of 0 for VF vscale x 8 For instruction:   %indvars.iv = phi i64 [ 0, %for.cond1.preheader.us ], [ %indvars.iv.next, %for.body4.us ]
LV: Found an estimated cost of 0 for VF vscale x 8 For instruction:   %arrayidx.us = getelementptr inbounds i8, ptr %src1.addr.031.us, i64 %indvars.iv
LV: Found an estimated cost of 1 for VF vscale x 8 For instruction:   %0 = load i8, ptr %arrayidx.us, align 1, !tbaa !6
LV: Found an estimated cost of 0 for VF vscale x 8 For instruction:   %conv.us = zext i8 %0 to i16
LV: Found an estimated cost of 0 for VF vscale x 8 For instruction:   %arrayidx6.us = getelementptr inbounds i8, ptr %src2.addr.030.us, i64 %indvars.iv
LV: Found an estimated cost of 1 for VF vscale x 8 For instruction:   %1 = load i8, ptr %arrayidx6.us, align 1, !tbaa !6
LV: Found an estimated cost of 0 for VF vscale x 8 For instruction:   %conv7.us = zext i8 %1 to i16
LV: Found an estimated cost of 1 for VF vscale x 8 For instruction:   %add.us = add nuw nsw i16 %conv.us, 1
LV: Found an estimated cost of 1 for VF vscale x 8 For instruction:   %add8.us = add nuw nsw i16 %add.us, %conv7.us
LV: Found an estimated cost of 1 for VF vscale x 8 For instruction:   %shr.us = lshr i16 %add8.us, 1
LV: Found an estimated cost of 0 for VF vscale x 8 For instruction:   %conv9.us = trunc i16 %shr.us to i8
LV: Found an estimated cost of 0 for VF vscale x 8 For instruction:   %arrayidx11.us = getelementptr inbounds i8, ptr %dst.addr.032.us, i64 %indvars.iv
LV: Found an estimated cost of 1 for VF vscale x 8 For instruction:   store i8 %conv9.us, ptr %arrayidx11.us, align 1, !tbaa !6
LV: Found an estimated cost of 1 for VF vscale x 8 For instruction:   %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
LV: Found an estimated cost of 1 for VF vscale x 8 For instruction:   %exitcond.not = icmp eq i64 %indvars.iv.next, %wide.trip.count
LV: Found an estimated cost of 0 for VF vscale x 8 For instruction:   br i1 %exitcond.not, label %for.cond1.for.cond.cleanup3_crit_edge.us, label %for.body4.us, !llvm.loop !9
LV: Vector loop of width vscale x 8 costs: 0 (assuming a minimum vscale of 2).
LV: Found an estimated cost of 0 for VF vscale x 16 For instruction:   %indvars.iv = phi i64 [ 0, %for.cond1.preheader.us ], [ %indvars.iv.next, %for.body4.us ]
LV: Found an estimated cost of 0 for VF vscale x 16 For instruction:   %arrayidx.us = getelementptr inbounds i8, ptr %src1.addr.031.us, i64 %indvars.iv
LV: Found an estimated cost of 1 for VF vscale x 16 For instruction:   %0 = load i8, ptr %arrayidx.us, align 1, !tbaa !6
LV: Found an estimated cost of 2 for VF vscale x 16 For instruction:   %conv.us = zext i8 %0 to i16
LV: Found an estimated cost of 0 for VF vscale x 16 For instruction:   %arrayidx6.us = getelementptr inbounds i8, ptr %src2.addr.030.us, i64 %indvars.iv
LV: Found an estimated cost of 1 for VF vscale x 16 For instruction:   %1 = load i8, ptr %arrayidx6.us, align 1, !tbaa !6
LV: Found an estimated cost of 2 for VF vscale x 16 For instruction:   %conv7.us = zext i8 %1 to i16
LV: Found an estimated cost of 2 for VF vscale x 16 For instruction:   %add.us = add nuw nsw i16 %conv.us, 1
LV: Found an estimated cost of 2 for VF vscale x 16 For instruction:   %add8.us = add nuw nsw i16 %add.us, %conv7.us
LV: Found an estimated cost of 2 for VF vscale x 16 For instruction:   %shr.us = lshr i16 %add8.us, 1
LV: Found an estimated cost of 1 for VF vscale x 16 For instruction:   %conv9.us = trunc i16 %shr.us to i8
LV: Found an estimated cost of 0 for VF vscale x 16 For instruction:   %arrayidx11.us = getelementptr inbounds i8, ptr %dst.addr.032.us, i64 %indvars.iv
LV: Found an estimated cost of 1 for VF vscale x 16 For instruction:   store i8 %conv9.us, ptr %arrayidx11.us, align 1, !tbaa !6
LV: Found an estimated cost of 1 for VF vscale x 16 For instruction:   %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
LV: Found an estimated cost of 1 for VF vscale x 16 For instruction:   %exitcond.not = icmp eq i64 %indvars.iv.next, %wide.trip.count
LV: Found an estimated cost of 0 for VF vscale x 16 For instruction:   br i1 %exitcond.not, label %for.cond1.for.cond.cleanup3_crit_edge.us, label %for.body4.us, !llvm.loop !9
LV: Vector loop of width vscale x 16 costs: 0 (assuming a minimum vscale of 2).
```

**Comment by vfdff** [comment]
fixed with [D157628](https://reviews.llvm.org/D157628)
